subsys/net/lib/coap/coap.c

```diff
@@ -1817,23 +1819,44 @@ struct coap_reply *coap_response_received(
 		return NULL;
 	}
```

Comment:
```
	if (!is_empty_message(response) && coap_packet_is_request(response)) {
		/* Request can't be response */
		return NULL;
	}
```
The **!** before `is_empty_message` is wrong. Both requests and empty messages can't be responses, see table below.

Reply:
Good point, that check indeed seems wrong. Changed to `if (is_empty_message(response) || coap_packet_is_request(response))` and added test cases to cover empty reply and request case.